numba/cuda/tests/cudapy/test_fastmath.py

```diff
@@ -136,24 +171,6 @@ def f10(r, x, y):
         except ZeroDivisionError:
             self.fail("Divide in fastmath should not throw ZeroDivisionError")
 
-    def test_device(self):
-        # fastmath option is ignored for device function
-        @cuda.jit("float32(float32, float32)", device=True)
-        def foo(a, b):
-            return a / b
-
-        def bar(arr, val):
-            i = cuda.grid(1)
-            if i < arr.size:
-                arr[i] = foo(i, val)
-
-        sig = (float32[::1], float32)
-        fastver = cuda.jit(sig, fastmath=True)(bar)
-        precver = cuda.jit(sig)(bar)
-
-        self.assertIn('div.full.ftz.f32', fastver.ptx[sig])
-        self.assertNotIn('div.full.ftz.f32', precver.ptx[sig])
-
```

Comment:
I think this test tested something a little different to all the other test cases in this file - the propagation of fastmath from callers into callees. Presently this doesn't actually work - compiling `bar` with fastmath doesn't result in the same code for `foo` as if it were compiled with fastmath, but it should. 

I think it might be worth reinstating this test case, in a slightly different form:

```python
    @unittest.expectedFailure     
    def test_device_fastmath_propagation(self):    
        # The fastmath option doesn't presently propagate to device functions     
        # from their callees - arguably it should do, so this test is presently     
        # an xfail.     
        @cuda.jit("float32(float32, float32)", device=True)    
        def foo(a, b):    
            return a / b    
      
        def bar(arr, val):    
            i = cuda.grid(1)    
            if i < arr.size:    
                arr[i] = foo(i, val)    
            
        sig = (float32[::1], float32)    
        fastver = cuda.jit(sig, fastmath=True)(bar)    
        precver = cuda.jit(sig)(bar)    
            
        # Variants of the div instruction are further documented at:     
        # https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#floating-point-instructions-div     
    
        # The fast version should use the "fast, approximate divide" variant     
        self.assertIn('div.approx.f32', fastver.ptx[sig])    
        # The precise version should use the "IEEE 754 compliant rounding"     
        # variant, and neither of the "approximate divide" variants.     
        self.assertIn('div.rn.f32', precver.ptx[sig])    
        self.assertNotIn('div.approx.f32', precver.ptx[sig])    
        self.assertNotIn('div.full.f32', precver.ptx[sig])
```

The test case previously ensured our behaviour was consistent with what it was in the past, but I think it's more appropriate to mark this as an expected failure because we know the present behaviour is wrong. However, when changes to the propagation of flags is made (potentially to the core of Numba rather than specific to the CUDA target) it will be good to detect that this gets fixed.

Reply:
> the propagation of fastmath from callers into callees

Ah ok, I didn't realize the test was to test the propagation property. It seems to me propagation should be configurable. But I'd like to investigate more about how much granularity does nvcc allows user to configure different compilation unit with the fast-math flag and Numba should provide similar level of granularity.

Reply:
There's an issue for some general discussion around the propagation of flags here: https://github.com/numba/numba/issues/7168